File: java/dsa/algorithms/sliding-window/min-window.java
public class Solution {
    public String minWindow(String s, String t) {
        if (s.length() < t.length()) return "";

        Map<Character, Integer> targetFreq = new HashMap<>();
        for (char c : t.toCharArray()) {
            targetFreq.put(c, targetFreq.getOrDefault(c, 0) + 1);
        }

        int requiredMatches = targetFreq.size();
        Map<Character, Integer> windowFreq = new HashMap<>();
        int left = 0, matchCount = 0;
        int minLength = Integer.MAX_VALUE, bestStart = 0;

        for (int right = 0; right < s.length(); right++) {
            // 1️⃣ Expand
            char c = s.charAt(right);
            if (targetFreq.containsKey(c)) {
                windowFreq.put(c, windowFreq.getOrDefault(c, 0) + 1);
                if (windowFreq.get(c).equals(targetFreq.get(c))) {
                    matchCount++;
                }
            }

            // 2️⃣ Condition: while valid
            while (matchCount == requiredMatches) {
                // 4️⃣ Record / Optimize
                if (right - left + 1 < minLength) {
                    minLength = right - left + 1;
                    bestStart = left;
                }

                // 3️⃣ Contract
                char leftChar = s.charAt(left);
                if (targetFreq.containsKey(leftChar)) {
                    if (windowFreq.get(leftChar).equals(targetFreq.get(leftChar))) {
                        matchCount--;
                    }
                    windowFreq.put(leftChar, windowFreq.get(leftChar) - 1);
                }
                left++;
            }
        }

        return minLength == Integer.MAX_VALUE ? "" : s.substring(bestStart, bestStart + minLength);
    }
}

File: javascript/dsa/algorithms/sliding-window/min-window.js
function minWindow(s, t) {
    if (s.length < t.length) return "";

    const targetFreq = {};
    for (const char of t) {
        targetFreq[char] = (targetFreq[char] || 0) + 1;
    }

    const requiredMatches = Object.keys(targetFreq).length;
    const windowFreq = {};
    let left = 0, matchCount = 0;
    let minLength = Infinity, bestStart = 0;

    for (let right = 0; right < s.length; right++) {
        // 1️⃣ Expand
        const char = s[right];
        if (targetFreq[char]) {
            windowFreq[char] = (windowFreq[char] || 0) + 1;
            if (windowFreq[char] === targetFreq[char]) {
                matchCount++;
            }
        }

        // 2️⃣ Condition: while valid
        while (matchCount === requiredMatches) {
            // 4️⃣ Record / Optimize
            if (right - left + 1 < minLength) {
                minLength = right - left + 1;
                bestStart = left;
            }

            // 3️⃣ Contract
            const leftChar = s[left];
            if (targetFreq[leftChar]) {
                if (windowFreq[leftChar] === targetFreq[leftChar]) {
                    matchCount--;
                }
                windowFreq[leftChar]--;
            }
            left++;
        }
    }

    return minLength === Infinity ? "" : s.substring(bestStart, bestStart + minLength);
}